numpy/core/src/multiarray/ctors.c

```diff
@@ -1090,6 +1090,14 @@ PyArray_NewFromDescr_int(
     }
 
     fa = (PyArrayObject_fields *) subtype->tp_alloc(subtype, 0);
+    if (!PyDataType_REFCHK(descr) && subtype == &PyArray_Type && data == NULL) {
+        /*
+         * tp_alloc will have enabled GC tracking, but it's only necessary
+         * for object arrays. Subtypes may also define additional fields, so
+         * a subtype would have to handle this by itself.
+         */
+        PyObject_GC_UnTrack(fa);
```

Comment:
Are you sure this is correct? What does cpython do about `dict` subclasses, for instance?

Reply:
To be honest, I have no thought to much about the untracking, it is the method I took from the original PR. But I am not sure how you think this can be incorrect?

Reply:
OK, I actually looked a bit at the Python code, and it seems to me that this is _not_ correct and I would have expected it to run into asserts on a debug python build (under the assumption that such an array is added to container that implements cyclic gc). I am a bit surprised I did not notice that (although without checking I am not 100% we have CI which runs a python debug build).

There would be solution to this: One can implement `tp_is_gc` which returns `False` after untracking/for arrays which are not added to the GC. (There is probably only a single use case of this: Python uses it to distinguish between HEAP and static types.)

Right now, I think this is arguably a bug, and chances are that having a few more arrays does not matter (there must be a huge number of other python containers around that need to do more traversing).

Still planning to try out the speed penalty for keeping a huge object array around, but even if it is noticable, I am not sure it matters too much as to whether or not we should do this.

Reply:
Why is the object being tracked at this point already? From the docs:
```
PyObject_GC_Track` should be called once all the fields followed by the tp_traverse handler become valid, usually near the end of the constructor.
```
Could we just call `PyObject_GC_Track` here instead?

Reply:
In fact, aren't we supposed to call `PyObject_GC_Track` ourselves anyway? I'm not seeing the bit of the docs that imply its called for us.

Reply:
The default `tp_alloc` calls it for you when `Py_TPFLAGS_HAVE_GC`. Strictly speaking: probably yes. But I also do not think it matters, since there we own the GIL the whole time here, and there is no call back into python before everything is finished up.

Anyway, if someone wants to review+merge this, I am happy to look at this, otherwise, for now we can also close until someone actually complains due to a memory leak. The new changes in python probably need a bit of cleaning, and the release notes may need to warn hypothetical hacks that subclass from C at least.